### PR TITLE
Sidam 3596 bug health check app

### PR DESCRIPTION
**Pull request by matteoparussini**
- Checking probe status only when MARK_AS_DOWN otherwise status is ok
- Adding log warn and simplify method complexity
- update lib version

### JIRA link (if applicable) ###

[SIDM-3596](https://tools.hmcts.net/jira/browse/SIDM-3596)

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java [hunk 1]
```diff
@@ -29,16 +29,22 @@ public ScheduledHealthProbeIndicator(
...
             return this.healthProbe.probe();
         }
-        return status == Status.UP
-                && LocalDateTime.now(clock).isBefore(statusDateTime.plus(freshnessInterval, ChronoUnit.MILLIS));
+
+        if (failureHandling == HealthProbeFailureHandling.MARK_AS_DOWN){
+            return status == Status.UP
+                    && LocalDateTime.now(clock).isBefore(statusDateTime.plus(freshnessInterval, ChronoUnit.MILLIS));
+        } else {
+            log.warn("{}: status evaluation ignored for this type of probe. failureHandling: {}", this.healthProbe.getName(), failureHandling.toString());
```
Comment (kremi) [inline src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: Could we please also log the current `status` value?

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java [hunk 2]
```diff
@@ -29,16 +29,22 @@ public ScheduledHealthProbeIndicator(
...
+            return true;
```
Comment (kremi) [inline src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: I think it's worth a testing for this case.

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java [hunk 3]
```diff
@@ -51,20 +57,25 @@ protected void refresh() {
...
-                }
-            }
+            printLogMessage(probeResult);
+
             this.status = probeResult ? Status.UP : Status.DOWN;
             this.statusDateTime = LocalDateTime.now(clock);
         } else if (failureHandling == HealthProbeFailureHandling.IGNORE) {
             log.warn("{}: DOWN state ignored", this.healthProbe.getName());
         }
     }
```
Comment (nikola-naydenov-hmcts) [inline src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: minor: if you make this method return the new status, you could greatly simplify the code here
i.e Status newStatus = probeResult ? Status.UP : Status.DOWN;
(if logInfo and status != this.status) print message;
return newStatus;

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java (hunk 1 above)
Comment (matteoparussini) [reply src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: ok

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java (hunk 2 above)
Comment (matteoparussini) [reply src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: @kremi, new test was added

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java [hunk 4]
```diff
@@ -29,16 +29,21 @@ public ScheduledHealthProbeIndicator(
...
-        taskScheduler.scheduleWithFixedDelay(() -> refresh(), checkInterval);
+        taskScheduler.scheduleWithFixedDelay(this::refresh, checkInterval);
     }
 
     @Override
     public boolean isOkay() {
-        if (status == Status.UNKNOWN) {
-            return this.healthProbe.probe();
+        if (failureHandling == HealthProbeFailureHandling.MARK_AS_DOWN){
+            if (status == Status.UNKNOWN) {
```
Comment (kremi) [inline src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: I think we need to evaluate the probe in any case (irrelevant of the failureHandling) if the status is unknown.

File: src/test/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicatorTest.java [hunk 5]
```diff
@@ -105,4 +105,12 @@ public void testIsOkay_ignoreDown() {
         assertThat(ignoringScheduledHealthProbe.isOkay(), is(true));
         verify(healthProbe, times(2)).probe();
     }
+
+    @Test
+    public void testIsOkay_alwaysTrueWhenIgnoreProbe() {
```
Comment (kremi) [inline src/test/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicatorTest.java]: This is a good test but it does not cover the scenario that was failing in our latest incident: should be `testIsOkay_successWhenNotFreshAndIgnoreDown`

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java (hunk 4 above)
Comment (matteoparussini) [reply src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: My understanding was that when the HealthProbeFailureHandling is set as IGNORE, the method isOkay should always return true, regardless the state. If it's not like this, I'll roll back as it was before.

Comment (kremi) [reply src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: You are right but we should run the probe anyway. Maybe:
```
if (status == Status.UNKNOWN) {
  return this.healthProbe.probe() || failureHandling == HealthProbeFailureHandling.IGNORE;
}
```
Running the probe might produce some logs which help with troubleshooting (if required).

Comment (matteoparussini) [reply src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: OK, I understand. I'll update it

File: src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java (hunk 3 above)
Comment (matteoparussini) [reply src/main/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicator.java]: Returning a status in a method that is called "printLogMessage" it sounds misleading. The logging depends on the status change, so comparing newStatus and oldStatus it makes sense. I'll update it.

File: src/test/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicatorTest.java (hunk 5 above)
Comment (matteoparussini) [reply src/test/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicatorTest.java]: The problem with that test is that when a Probe is IGNORE and its probe check is false, at the moment, we are not updating the status to DOWN (we are just printing it). Is that correct?

Comment (kremi) [reply src/test/java/uk/gov/hmcts/reform/idam/health/probe/ScheduledHealthProbeIndicatorTest.java]: Yeah, that's true - sorry I've misspelled the name of the test, should be `testIsOkay_successWhenNotFreshAndIgnore` (isOkay() returns `true` even when `LocalDateTime.now(clock).isBefore(statusDateTime.plus(freshnessInterval, ChronoUnit.MILLIS));`

**Pull request merged**